internal/commandhandler/manager.go

```diff
@@ -0,0 +1,70 @@
+package commandhandler
+
+import (
+	"errors"
+	"sync"
+	"sync/atomic"
+
+	"github.com/dicedb/dice/internal/shard"
+)
+
+type Manager struct {
+	activeCmdHandlers sync.Map
+	numCmdHandlers    atomic.Int32
+	maxCmdHandlers    int32
+	ShardManager      *shard.ShardManager
+	mu                sync.Mutex
+}
+
+var (
+	ErrMaxCmdHandlersReached = errors.New("maximum number of command handlers reached")
+	ErrCmdHandlerNotFound    = errors.New("command handler not found")
+)
+
+func NewManager(maxCmdHandlers int32, sm *shard.ShardManager) *Manager {
+	return &Manager{
+		maxCmdHandlers: maxCmdHandlers,
+		ShardManager:   sm,
+	}
+}
+
+func (m *Manager) RegisterCommandHandler(cmdHandler *BaseCommandHandler) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if m.CommandHandlerCount() >= m.maxCmdHandlers {
+		return ErrMaxCmdHandlersReached
+	}
+
+	responseChan := cmdHandler.responseChan
+	preprocessingChan := cmdHandler.preprocessingChan
+
+	if responseChan != nil && preprocessingChan != nil {
+		m.ShardManager.RegisterCommandHandler(cmdHandler.ID(), responseChan, preprocessingChan) // TODO: Change responseChan type to ShardResponse
+	} else if responseChan != nil && preprocessingChan == nil {
+		m.ShardManager.RegisterCommandHandler(cmdHandler.ID(), responseChan, nil)
+	}
```

Comment:
maybe this can be placed before this if else condition and each of the two channels can be checked and initialized separately. It would avoid errors where either one of the two channels is somehow uninitialized and doesn't fit into the existing if-else logic.

Reply:
@JyotinderSingh - One clarification. Currently `AcceptConnectionRequests` is the only one using `RegisterCommandHandler` function and it is initializing both channels always. I am not sure why this check exist in the first place.

Reply:
I removed the entire if else check and added a check on respChan. Please see this [comment](https://github.com/DiceDB/dice/pull/1358#discussion_r1876394434)